docs/readmes/proposals/p021_mme_migrate_to_c++.md

```diff
@@ -0,0 +1,143 @@
+---
+id: p021_mme_migrate_to_c++
+title: C++ Migration
+hide_title: true
+---
+
+
+# Proposal: C++ Migration
+
+Author(s): [@alexrod, @electronjoe, @kozat, @lionelgo]
+
+Last updated: 09/15/2021
+
+Discussion at
+[https://github.com/magma/magma/issues/4074](https://github.com/magma/magma/issues/4074).
+
+## Abstract
+
+This document describes the steps to migrate progressively, from C to C++, the MME core.
+
+## Background
+
+Initially and historically, the development of MME magma was done only in C.
+
+Many third party libraries and features that were integrated and or developed around the MME core are written in C++.
+
+We are now to the point where, for an easier and cleaner way of integrating new C++ libraries, for the ease of the maintenance of the code, a C to C++ migration seems unavoidable.
+
+## Proposal
+
+The proposal is constrained by a lack of performance of the current context serialization library (protobuf).
+ 
+Because the replacement libraries may have a big impact on the MME core code, this library migration will be studied first.
+
+
+### State Serialization
+The serialization performance of protobuf is insufficient, so a serialization library that allows better performance must be selected.
+
+The bad performance is due to the global serialization process (translation, memory allocation, copy).
+
+The replacement libraries considered are "cap'n proto" and "Flat buffers”, they support C/C++ standard API, binary output format and do not need memory copy (zero copy).
```

Comment:
the way cap'n proto and flat buffers achieve 0copy is making their structures read only. this yields great performance benefit during read operations. we do not do many state deserialization/reads in the MME codebase.

once mutation of state and then re-serialization is necessary, cap'n proto and flat buffers loses most of its performance advantage.

while our current pattern of hashtable -> protobuf obj -> serialization yields a lot of wasted cycles, storing the state in protobuf objs directly will alleviate most of this (specifically, a dramatic decrease in heap alloc/free). going from an in-memory protobuf -> serialized bytes should be close to the same performance of cap'n proto and flat buffers.

sticking with protobuf if possible is a significantly superior strategy. protobuf has dramatically more support compared to cap'n proto and flat buffers. it also matches model/codegen and ser/des choices in grpc and other services working with MME. while one could consider MME state serialization to be a "private" concern, it's still a slight advantage for all other sections of the code to be able to share models and make dev tooling universal.

Reply:
> going from an in-memory protobuf -> serialized bytes should be close to the same performance of cap'n proto and flat buffers.

I don't buy this without some data - protobuf is not stored in memory compressed is it? That would be a serious penalty for any accesses.  So serialization involves all the encoding overhead of Protobuf (which is substantial - given the objective is to minimize size on the wire).

> sticking with protobuf if possible is a significantly superior strategy.

Agreed. Should only modify the strategy if the performance impact (system wide, not MME specifically) is enabling / make or break.

Reply:
Here is direct data.

I believe we would be comparing `FlatBuffers (binary)` with `Protocol Buffers Lite` (actually not Lite for us, presumably worse for full fledge Protobuf).  Encoding speed for 1 million encodes in the FlatBuffers benchmark (by Google) yields **~60x speedup for FlatBuffers vs PB Lite**.

I went to look for their source code for the benchmark to confirm that they were not measuring the population time of the fields in the buffers, just encoding time - but the benchmarks are long gone and I've not rediscovered them in binary search of the git repo. Still - the encoding effort delta makes sense to me (even once in memory) - Protobuf does a lot of work to encode with minimal size.

Reply:
> [Here](https://google.github.io/flatbuffers/flatbuffers_benchmarks.html) is direct data from Google.
> 
> I believe we would be comparing `FlatBuffers (binary)` with `Protocol Buffers Lite` (actually not Lite for us, presumably worse for full fledge Protobuf). Encoding speed for 1 million encodes in the FlatBuffers benchmark (by Google) yields **~60x speedup for FlatBuffers vs PB Lite**.
> 
> I went to look for their source code for the benchmark to confirm that they were not measuring the population time of the fields in the buffers, just encoding time - but the benchmarks are long gone and I've not rediscovered them in binary search of the git repo. Still - the encoding effort delta makes sense to me (even once in memory) - Protobuf does a lot of work to encode with minimal size.

This benchmark is not representative of our scenario: update + encode.

Comparing pre-canned immutable flatbuf serialization to mutable protobuf serialization is not representative of our use case. Here are the scenarios to consider:

- State stored directly in mutable protobuf objs, update field value in protobuf obj (essentially free), serialize out
- State stored directly in flatbuf, rebuilding flatbuf on each update, serialize out
- State stored in hashtable (or other datastructure), convert to flatbuf (this part is effectively the same as serializing), serialize out

1. Doing a straight flatbuf vs protobuf serialization benchmark is a waste of time, we already know the results.
2. Replacing the current convert to protobuf + serialize with convert to flatbuf + serialize will yield better performance. However, it will introduce a new dependency and be hard to migrate away from.
3. Replacing current, in-memory state representation w/ protobuf is the best case if it provides acceptable performance. I don't want us to get stuck on step 2; if we want to benchmark things, then let's compare 2 vs. 3, not 1 vs. 2.

Reply:
Comparing 2 vs 3 sounds good - and agree that a reasonable first step is in memory protobuf representation / mutation in place.  Removes a whole lot of non-proto hash map iteration and copies (which may be plenty to resolve the performance bottleneck.

Reply:
> we do not do many state de-serialization/reads in the MME code base.
Actually I agree, depends on constraints and strategy and deployments envisioned.
If you have tight constraints on occupied memory, you could flush your context to persistent storage and free it after the received MODIFY_BEARER_RESPONSE or with less aggressiveness when releasing UE_CONTEXT based on number of UE attached threshold.
If you want to deploy magma in a cluster you may become even more aggressive by choosing to follow the stateless paradigm.